snuba/web/query.py

```diff
@@ -201,26 +233,22 @@ def log_query_and_update_stats(
     timer: Timer,
     stats: MutableMapping[str, Any],
     status: str,
+    query_list: List[ClickhouseQueryMetadata],
     query_settings: Mapping[str, Any],
 ) -> MutableMapping:
     """
     If query logging is enabled then logs details about the query and its status, as
     well as timing information.
     Also updates stats with any relevant information and returns the updated dict.
     """
-
     stats.update(query_settings)
+
     if settings.RECORD_QUERIES:
-        # send to redis
-        state.record_query(
-            {
-                "request": request.body,
-                "sql": sql,
-                "timing": timer,
-                "stats": stats,
-                "status": status,
-            }
+
+        query_metadata = ClickhouseQueryMetadata(
+            request=request.body, sql=sql, timer=timer, stats=stats, status=status
```

Comment:
Since `Timer` is mutable (and mutated when `mark` is called), each passing the timer instance here is going to result in all of the metadata objects containing the timing data for _all_ of the queries when `record_query` is called.

Similarly, I know that the request body is mutated as part of split query, but I don't know if that is going to result in this assignment being similarly unsafe — probably good to double check, unless you have already done that. (Even if it does work right now it might be better to be cautious and copy it anyway, since I don't think it is reasonable for anybody working on that function to have the foresight to anticipate how mutating the request would effect query logging.)